### ISSUE TITLE
Use dedicated pools for publishing jobs

### ISSUE THREAD

**Issue by mmitche** [dotnet/dnceng#1145]
We should have a dedicated pool for the publishing jobs/stages:
- Publish To Build Asset Registry
- Publish Using Darc
- Maestro promotion job

The idea would be to reduce the wait time for machines to do this work, ideally to 0. These jobs are on the critical path of build completion, and we stand to gain perhaps 2 hours on an end to end build time for .NET 8/7/6. We can use a separate pool for these jobs. These jobs are typically short (2-12 minutes) and we average only a few hundred hours of total m,achine time/month.

After the pool is available, we would apply this pool to the arcade templates as well as the maestro promotion job.

### Release Note Category

- [x] Feature changes/additions 
- [ ] Bug fixes
- [ ] Internal Infrastructure Improvements

### Release Note Description

Created new dedicated pools for publishing jobs: `NetCore1ESPool-Publishing-Internal`.

Due to changes in how we addressed the related issue dotnet/arcade#14445, we took a breaking change within the Maestro++ infrastructure. Please see dotnet/arcade-services#3258 for our reaction to that break (in `darc`).

**Comment by riarenas** [comment]
This pool might be a good candidate to enable [Spot VMs](https://eng.ms/docs/cloud-ai-platform/devdiv/one-engineering-system-1es/1es-docs/1es-hosted-azure-devops-pools/spot-vm) as the publishing jobs are short.

We can also use the [available pre-provisioning options](https://eng.ms/docs/cloud-ai-platform/devdiv/one-engineering-system-1es/1es-docs/1es-hosted-azure-devops-pools/agent-creation-performance) that we can't take advantage of for the larger machine pools.

**Comment by dougbu** [comment]
- Is this about a new `pool:` in the Azure pipeline sense or a few `OneESHostedPoolImage` images for `ImageOverride` use❔
- If it's about 1ES images, which specific ones are needed❔ I see windows.vs2019.amd64 used everywhere today
- Should something about the "new" thing(s) be locked down so that e.g., only the VMR pipeline can use it / them❔
- Will the arcade job and post-build updates happen only for servicing branches❔

**Comment by tkapin** [comment]
Is this getting prioritized? We need to know when it's going to be rolled out so we can reason about how this affects the release times.

**Comment by premun** [comment]
>- Should something about the "new" thing(s) be locked down so that e.g., only the VMR pipeline can use it / them❔

The idea is that all official repo pipelines will use this for the darc publishing stages.
Also the release/staging pipeline would use these.
Possibly also the code-mirror and other infrastructural pipelines?

**Comment by AlitzelMendez** [comment]
@mmitche do you have any insights in which should be the Maximum pool size of this pool?

**Comment by mmitche** [comment]
> @mmitche do you have any insights in which should be the Maximum pool size of this pool?

Not sure on maximum. Probably not more than 10? When I calculated a minimum it didn't come out to much more than 4-6.

**Comment by mmitche** [comment]
I tried to figure out how many publishing jobs ran simultaneously.

**Comment by AlitzelMendez** [comment]
Thanks @mmitche, I will create it with 10, and if it is not enough we can increase it 😄 

@tkapin, I just started working on the new pool, so it would take at least a week for this to rollout (the new pool) and after that someone will continue the work of updating the templates (which is when this will actually be visible, does this work with you?

**Comment by AlitzelMendez** [comment]
I talked with Ricardo a little bit more about this and he brough out the topic of having a stateful machines, this will help us to always have machines available, and considering that we don't need a ton of machines this could be really beneficial, there is a little of information in here: [FAQ | 1ES On EngHub](https://eng.ms/docs/cloud-ai-platform/devdiv/one-engineering-system-1es/1es-docs/1es-hosted-azure-devops-pools/faq#stateful-vs-stateless)

One of the things is that we will need to define how often a machine is recycled, the default is 7 days, but I am not that familiar with the jobs to know if this could be a lot of time and can become problematic. 

All this was brough of due to the concern that if we create just a regular stateless pool, there are chances that the problem doesn't get actually fixed as we will spend a lot of time creating clean machines which is not necessary.

does anyone have thoughts about this?

**Comment by dougbu** [comment]
what's the advantage of a stateless pool over a scale set w/ a non-zero minimum number of instances❔

**Comment by riarenas** [comment]
I think the FAQ does a better job here than I could: https://eng.ms/docs/cloud-ai-platform/devdiv/one-engineering-system-1es/1es-docs/1es-hosted-azure-devops-pools/faq

**Comment by riarenas** [comment]
https://eng.ms/docs/cloud-ai-platform/devdiv/one-engineering-system-1es/1es-docs/1es-hosted-azure-devops-pools/agent-creation-performance, which is linked in previous comments is also still relevant.

**Comment by dougbu** [comment]
feels like stateless has the potential to cause problems w/ reuse of the _same_ workspace on the agents as well as potential disk space issues relatively quickly (given e.g., post-build downloads source from the repo,  not arcade). instead, choosing `Automatic` for `Agent Performance` / 
![image](https://github.com/dotnet/dnceng/assets/6431421/f77a4a02-c313-4799-95c7-40b91de38be8)
may be predictive enough to have a significant benefit while avoiding agent reuse. if @riarenas and @mmitche agree, I (or my replacement next week) can make the manual changes in one of our staging pools, see how the JSON changes, copy that into Ali's PR, then revert the staging changes

**Comment by dougbu** [comment]
fix is in [!36293](https://dev.azure.com/dnceng/internal/_git/dotnet-helix-machines/pullrequest/36293) if we decide the existing approach is fine

**Comment by dougbu** [comment]
I take that back: PR validation doesn't actually exercise `Deploy1ESHostedPools` will need to run a manual build against Ali's branch to confirm new pool is created successfully

**Comment by riarenas** [comment]
My strong preference is for this pool to always have machines available, or to get some concrete numbers that tell us that it'll be an improvement to just isolate the machines from the rest of the pools while keeping the 20min provisioning time of windows agents.

**Comment by dougbu** [comment]
I completely agree w/ always available agents. auto mode should do that simply because publishing happens all the time. might have different thoughts if we were doing this only in servicing branches. are we❔

what telemetry do we already have for the experiment❔ I'm familiar w/ work item queue and start times. do we have similar info about build jobs❔

**Comment by dougbu** [comment]
so far I haven't done anything that registers in the JSON view. it's possible that'll change overnight.

----

we don't get any traffic in staging. what should I do to get some telemetry and cost data❔

----

speaking of costs, I suggest disabling all images in this pool except `windows.vs2019.amd64` and maybe `windows.vs2022.amd64` or `1es-windows-2022` to keep costs somewhat contained. could add `build.ubuntu.2004.amd64` and a couple more for the VMR but let's watch costs first. agreed❔

**Comment by premun** [comment]
Are we expected to use this new pool in the staging/release pipelines? I'd say it's expected?

And if so, what did you land on yesterday? Will you need a list of images used in the release infra?

**Comment by dougbu** [comment]
> Are we expected to use this new pool in the staging/release pipelines? I'd say it's expected?

for now at least, the focus is entirely on the `Publish using Darc` and `Publish to Build Asset Registry` jobs in our arcade templates when executed in dnceng/internal. that means `windows.vs2019.amd64` is the sole available image.

we may expand this in the future, once we iron out the kinks and observe the costs. the promotion pipeline might (for example) be a good candidate

**Comment by dougbu** [comment]
status update: I created the necessary HelixStaging resources and am switching to testing a branch in dotnet-arcade to see if everything works end-to-end. I suspect the vnet or its subnet configuration will need some tweaks

**Comment by dougbu** [comment]
- builds [#20240125.6](https://dev.azure.com/dnceng/internal/_build/results?buildId=2362651&view=results) and [#20240125.7](https://dev.azure.com/dnceng/internal/_build/results?buildId=2362719&view=results) were successful w/ no default channel for my test branch
- after fixing a typo in the default channel setup, build [#20240126.2](https://dev.azure.com/dnceng/internal/_build/results?buildId=2363248&view=results) successfully published the assets to `General Testing` e.g., https://dev.azure.com/dnceng/public/_artifacts/feed/general-testing/NuGet/Microsoft.DotNet.Arcade.Sdk/overview/9.0.0-beta.24076.2

on to creating prod pool&hellip;

**Comment by dougbu** [comment]
actually, don't we need _both_ `NetCore1ESPool-Publishing-Internal` and `NetCore1ESPool-Publishing-Svc-Internal`❔

**Comment by dougbu** [comment]
so, packages got into General Testing just fine. but, is this sufficient validation to create the pool in production❔

for example, is there a way to test using the new pool in [Pipelines - Runs for Maestro Build Promotion (azure.com)](https://dev.azure.com/dnceng/internal/_build?definitionId=750&_a=summary) and should I❔

	
            Azure DevOps Services | Sign In

**Comment by dougbu** [comment]
in a more complete validation, I did 7 manual builds of my branch in the arcade-official-ci pipeline w/ no default channel selected. I then executed
``` powershell
darc add-build-to-channel --channel "General Testing" --no-wait --source-branch dougbu/try.new.pool --id {BAR id}
```
for all of the arcade builds. running agents in the pool changed approximately as I expected:
1. initially, 3 agents waited
    ![agents-before](https://github.com/dotnet/dnceng/assets/6431421/3ef177de-68e3-4d42-96b3-c0956e01dcbf)
2. a bit after the first three build promotions got agents, 7 more started up
   ![agents-starting](https://github.com/dotnet/dnceng/assets/6431421/b0bfcd8e-02ac-4b2c-9a12-c1e9d33e09b0)
   - this was slightly better than I expected b/c it covered both the immediate requests and the need to replace the 3 agents already in use
3. the new agents went through uninteresting states such as `provisioning` while the initial 3 finished their builds and were marked as `pendingReturn"
   ![agents-pending-return](https://github.com/dotnet/dnceng/assets/6431421/fd6d93dd-81c7-4383-a3b8-0e795f88f69f)
4. 4 of the new agents were allocated only slightly after the initial 3 were marked as `returned`
   ![agents-second-wave](https://github.com/dotnet/dnceng/assets/6431421/f72b0944-1d5c-410b-9ed7-32d142a13a13)
5. eventually, the pool was left with 3 `ready` agents. these were _not_ the initial 3
   ![agents-after](https://github.com/dotnet/dnceng/assets/6431421/2103db75-baf4-4ed4-9e32-a9691612601e)

**Comment by dougbu** [comment]
the promotion builds themselves were also pretty predictable
1. 3 builds started very quickly (after I remembered to allow the promotion pipeline to use the new pool
   ![runs-permission-granted](https://github.com/dotnet/dnceng/assets/6431421/762f2d91-39a4-444c-b6a4-7cb5f794955d)
2. about 6 minutes later the other 4 builds started
   ![runs-second-wave](https://github.com/dotnet/dnceng/assets/6431421/234a7584-1f2a-4250-b5a0-f42b2d9da68f)

**Comment by dougbu** [comment]
unfortunately the `D2_v5` agents ran much slower than our usual `D4a_v4` choice:
1. similar runs on `NetCore1ESPool-Internal` (our usual build promotion pool)
   ![runs-before-faster](https://github.com/dotnet/dnceng/assets/6431421/79f3e46f-7b79-42c6-87b5-00cec5c7c2b7)
2. the builds on `NetCore1ESPool-Publishing-Internal-Int` took about twice as long
   ![runs-after-slow](https://github.com/dotnet/dnceng/assets/6431421/efdde420-f1de-4b90-a1d2-713da782dafd)

**Comment by dougbu** [comment]
@ilyas1974 should I try `D4a_v4` again (cutting network speeds but going back to AMD64 and 16gb of memory), `D4as_v5` (maybe `D2as_v5` though that again has 8gb of memory), `D4_v5`, or something else❔

**Comment by dougbu** [comment]
note agent sizes didn't slow arcade-official-ci builds down. in fact, improved allocation speeds more than compensated for any running slowdown. 'main' in this pipeline has run 29 recent builds averaging `1h 15m 36s` duration while my branch did 14 runs lasting `1h 1m 30s`.

**Comment by ilyas1974** [comment]
If the d_v5 did not improve our speeds, my preference would be to try the d2a_v4 and see what that performance is like. D4a_v4 is what we currently use.

**Comment by dougbu** [comment]
with the updated size (`D2as_v5`) today's arcade-official-ci builds using the test pool averaged a bit longer (70.2 minutes instead of the 61.5 minutes I mentioned above). overall build times still better than the averages for 'main'. but, I haven't tested promotion build durations yet today &mdash; I tried to set up the default channel again but must have done something wrong.

----

why didn't the following command work this time❔
``` powershell
darc add-default-channel --repo https://dev.azure.com/dnceng/internal/_git/dotnet-arcade --channel 'General Testing' --branch refs/heads/dougbu/try.new.pool
```
that worked over the weekend (before I did a matching `delete-default-channel`) but not today, apparently.

in any case, that won't work end to end &mdash; Maestro Build Promotion should have used the regular pool for today's manual arcade builds. is there a way to have a regular build's publish jobs tell the promotion pipeline to use a different arcade branch❔ would be nice to watch the two build types overlapping and both using the same pool.

----

details&hellip;

oddly, the `Publish to Build Asset Registry` stage (e.g., in [build #20240129.10](https://dev.azure.com/dnceng/internal/_build/results?buildId=2365036&view=results)) shows
``` text
Metadata has been pushed. Build id in the Build Asset Registry is '210319'
Found the following default channels:
    https://dev.azure.com/dnceng/internal/_git/dotnet-arcade@dougbu/try.new.pool => (529) General Testing
Determined build will be added to the following channels: [529]
```
and that stage doesn't display a `Publish Using Darc` job but the `Publish Using Darc` job in the same-named stage shows
``` text
Build '210319' is already on all target channel(s).
```

**Comment by dougbu** [comment]
as I said said on Saturday, the Saturday promotions in the new pool ran for `6m 3s` to `7m 39s` minutes versus `2m 37s` to `2m 57s` minutes in the regular pool.  the promotions I just ran lasted `6m 20s` to `8m 20s` &mdash; a wider range and slightly slower than on Saturday. I suspect we're going in the wrong direction w/ machine sizes for the promotion pipeline. @mmitche should that pipeline (which all runs in a single job) be more CPU-bound than the publish steps❔

btw, pipeline duration analytics don't seem to show reasonable numbers compared to the build listings.

**Comment by mmitche** [comment]
@dougbu What failed there? I rarely use add-default-channel. You don't need one to get publishing, just "add-build-to-channel".

>  is there a way to have a regular build's publish jobs tell the promotion pipeline to use a different arcade branch❔ would be nice to watch the two build types overlapping and both using the same pool.

Yes, when you do `add-build-to-channel` with --source-branch <brnach name> it will use that arcade branch to do publishing. Otherwise, you can alter the yaml file in the build's branch to pass `--source-branch`

**Comment by mmitche** [comment]
A way to think of it is this way:
- The Publish Build Assets and Publish Using Darc jobs use what is specified in the repo's YAML file
- The promotion job uses what is specified in arcade's YAML (publish-v3.yml I think?). By default, arcade's main is used, `--source-branch`

**Comment by mmitche** [comment]
Some repos don't have `Publish Using Darc`. They have a `Publish` job which is the combination of the `Publish Using Darc` and `Publish To Build Asset Registry`. That's defined via `publishAssetsImmediately` in two places in the templates. e.g. https://github.com/dotnet/efcore/blob/main/azure-pipelines.yml#L316. It just allows for skipping an extra job if there is no Validate stage to run.

**Comment by mmitche** [comment]
> as I said said on Saturday, the Saturday promotions in the new pool ran for `6m 3s` to `7m 39s` minutes versus `2m 37s` to `2m 57s` minutes in the regular pool. the promotions I just ran lasted `6m 20s` to `8m 20s` — a wider range and slightly slower than on Saturday. I suspect we're going in the wrong direction w/ machine sizes for the promotion pipeline. @mmitche should that pipeline (which all runs in a single job) be more CPU-bound than the publish steps❔
> 
> btw, pipeline duration analytics don't seem to show reasonable numbers compared to the build listings.

I wouldn't have expected it, but it is extremely parallelized, so perhaps. Anyways, go with what is faster :)

**Comment by dougbu** [comment]
> @dougbu What failed there?

The arcade-official-ci builds I did yesterday didn't kick off promotion builds _at all_. It wasn't a huge deal b/c `add-build-to-channel` worked fine. I just wanted to test everything in one go.

> Anyways, go with what is faster :)

@ilyas1974 looking at the `D4{blah}` offerings, I believe the `D4as_v5` option may be the best choice here. Those machines are $85.7677/month using the one year savings plan versus $96.1045 for `D4a_v4` and `D4as_v4` SKUs (which have local temp storage) and $96.1191 for `D4_v5`, `D4s_v5`, `D4_v4`, and `D4s_v4` SKUs. I don't see much downside to using `D4as_v5` compared to the other offerings. Make sense❔

Side note: It seems `s` actually means a SKU supports premium and ultra SSD disks, not something of primary importance at this point.

**Comment by ilyas1974** [comment]
The only thing we need to consider is quota. By default, we get 100 cores within our subscription. Using the 4 core systems, that gives us ~25 concurrent systems. @mmitche, do you that will be enough or should we request more quota?

**Comment by dougbu** [comment]
> @mmitche, do you that will be enough or should we request more quota?

in https://github.com/dotnet/dnceng/issues/1145#issuecomment-1883908969, @mmitche said "Not sure on maximum. probably not more than 10?" but that was just for the publishing jobs. double that to include the promotion pipeline too and we're at 20. round up to 25 and we're fine 😀

**Comment by dougbu** [comment]
> The only thing we need to consider is quota.

I was also looking for validation that `D4as_v5` was a solid choice and didn't have downsides I missed.

**Comment by riarenas** [comment]
If this is using subnet delegation so that the VMs are crested in 1ES subscriptions, I'm not sure our quota is used here?

**Comment by dougbu** [comment]
> If this is using subnet delegation so that the VMs are crested in 1ES subscriptions, I'm not sure our quota is used here?

Core counts at least _seem_ unrelated to networking to me. I'm probably wrong&hellip;

**Comment by riarenas** [comment]
> If this is using subnet delegation so that the VMs are crested in 1ES subscriptions, I'm not sure our quota is used here?

Core counts at least seem unrelated to networking to me. I'm probably wrong…

It is and it isn't.

The subnet delegation is enabling a scenario where, 1ES will create the scale sets and VMs in their own managed subscriptions. My question is, since these scale sets are now in 1ES managed subscriptions instead of our own, do they consume our core quota? or 1ES' core quota? 

My guess is 1ES'

**Comment by dougbu** [comment]
looks like we're set on the SKU choice. the first fair test of arcade-official-ci w/ 9 builds using the `D4as_v5` SKU and waiting for promotion builds (that also used my branch) ran `1h 15m 36s` versus `1h 20m 39s` for 14 recent builds of main. that's not a huge difference but (a) the numbers come from pipeline analytics, which seem to over-estimate durations, (b) the Arcade build is quicker than some in dnceng/internal, and (c) Arcade doesn't produce that many assets for publication.

I don't understand why but the corresponding "Maestro Build Promotion" builds ran a minute or more longer than recent arcade promotions using the 'main' branch. I suggest we just use our normal `D4a_v4` default rather than `D4as_v5`. It may have something to do w/ local temp storage availability and reduced network I/O in that pipeline.

bottom line, I lean toward using the `D4a_v4` SKU even though it's slightly more expensive than what I last tested. it's core count anyhow, right @ilyas1974❔

----

I asked a separate question about using `Helix` versus one of our other subscriptions in Teams. fine to answer that here too.

**Comment by dougbu** [comment]
> After the pool is created, following wiki page should be updated: https://dev.azure.com/dnceng/internal/_wiki/wikis/DNCEng%20Services%20Wiki/554/Information-about-Build-Pools-in-the-'dnceng'-organization

done

**Comment by dougbu** [comment]
remaining work tracked in dotnet/arcade#14445, plus we likely need to automate deployment of `NetCore1ESPool-Publishing-Internal[-Int]` soon.
